src/lib.rs

```diff
@@ -59,20 +59,14 @@ macro_rules! impl_array {
             ) -> Result<(), Error> {
                 if let Some(next) = topic_parts.next() {
                     // Parse what should be the index value
-                    let i: usize = serde_json_core::from_str(next)?.0;
+                    let i: usize = serde_json_core::from_str(next).map_err(|_| Error::BadIndex)?.0;
```

Comment:
I have found it's simpler to use `.or(Err(Error::BadIndex))` instead of `map_err()` personally.